metadata/src/retain.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright 2019-2023 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+//! Utility functions to generate a subset of the metadata.
+
+use frame_metadata::{ExtrinsicMetadata, PalletMetadata, RuntimeMetadataV14, StorageEntryType};
+use scale_info::{form::PortableForm, interner::UntrackedSymbol};
+use std::{
+    any::TypeId,
+    collections::{BTreeMap, HashSet},
+};
+
+/// Collect all type IDs needed to represent the provided pallet.
+fn collect_pallet_types(pallet: &PalletMetadata<PortableForm>, type_ids: &mut HashSet<u32>) {
+    if let Some(storage) = &pallet.storage {
+        for entry in &storage.entries {
+            match entry.ty {
+                StorageEntryType::Plain(ty) => {
+                    type_ids.insert(ty.id);
+                }
+                StorageEntryType::Map { key, value, .. } => {
+                    type_ids.insert(key.id);
+                    type_ids.insert(value.id);
+                }
+            }
+        }
+    }
+
+    if let Some(calls) = &pallet.calls {
+        type_ids.insert(calls.ty.id);
+    }
+
+    if let Some(event) = &pallet.event {
+        type_ids.insert(event.ty.id);
+    }
+
+    for constant in &pallet.constants {
+        type_ids.insert(constant.ty.id);
+    }
+
+    if let Some(error) = &pallet.error {
+        type_ids.insert(error.ty.id);
+    }
+}
+
+/// Remove the generic type parameters of the type ID from the `PortableRegistry`
+/// of the metadata.
+fn remove_generic_type_params(metadata: &mut RuntimeMetadataV14, id: u32) {
+    let portable_ty = metadata
+        .types
+        .types
+        .get_mut(id as usize)
+        .unwrap_or_else(|| {
+            panic!("Metadata does not contain extrinsic type ID {id} registered; qed")
+        });
+
+    portable_ty.ty.type_params = Vec::new();
+}
+
+/// Collect all type IDs needed to represent the extrinsic metadata.
+///
+/// Ensure that the extrinsic types from the `PortableRegistry` are trimmed
+/// of the generic type parameters. Those are not needed and contain extra
+/// nested type IDs that increase unnecessarily the size of the generated metadata.
+fn collect_extrinsic_types(
+    extrinsic: &ExtrinsicMetadata<PortableForm>,
+    type_ids: &mut HashSet<u32>,
+) {
+    type_ids.insert(extrinsic.ty.id);
+
+    for signed in &extrinsic.signed_extensions {
+        type_ids.insert(signed.ty.id);
+        type_ids.insert(signed.additional_signed.id);
+    }
+}
+
+/// Update the given type using the new type ID from the portable registry.
+///
+/// # Panics
+///
+/// Panics if the [`scale_info::PortableRegistry`] did not retain all needed types.
+fn update_type(ty: &mut UntrackedSymbol<TypeId>, map_ids: &BTreeMap<u32, u32>) {
+    let old_id = ty.id;
+    let new_id = map_ids
+        .get(&old_id)
```

Comment:
you could also call `get(&old_id).copied().unwrap_or_else(..)` then remove replace `(*new_id).into()` with `new_id.into()`